### PR TITLE
feat(flow): render source/sink

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

Added `render_src/sink` functions to render source/sinks into dataflow, and node context to manage them

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- added `render_src/sink` to render source/sink in dataflow, using broadcast receiver in render_src, and unbounded_channel in render_sink

- added Node Context for helping transform sql to flow's execution plan
## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/src/transform.rs [hunk 1]
```diff
@@ -79,38 +92,32 @@ impl FunctionExtensions {
...
-        Ok((id, schema))
-    }
+/// To reuse existing code for parse sql, the sql is first parsed into a datafusion logical plan,
+/// then to a substrait plan, and finally to a flow plan.
+pub async fn sql_to_flow_plan(
+    ctx: &mut FlownodeContext,
+    engine: &Arc<dyn QueryEngine>,
+    sql: &str,
+) -> Result<TypedPlan, Error> {
+    let query_ctx = ctx.query_context.clone().unwrap();
```
Comment (zhongzc) [inline src/flow/src/transform.rs:102]: Is this operation safe?

File: src/flow/src/repr/relation.rs [hunk 2]
```diff
@@ -96,6 +106,49 @@ pub struct RelationType {
...
+        let old_to_new_col = BTreeMap::from_iter(
+            mfp.projection
+                .clone()
+                .into_iter()
+                .enumerate()
+                .map(|(new, old)| (old, new)),
+        );
+
+        // since it's just a mfp, we also try to preserve keys&time index information, if they survive mfp transform
+        let keys = self
```
Comment (zhongzc) [inline src/flow/src/repr/relation.rs:130]: I think it is necessary to explain the mapping rules between them

File: src/flow/src/repr/relation.rs [hunk 3]
```diff
@@ -96,6 +106,49 @@ pub struct RelationType {
...
+            .keys
+            .iter()
+            .filter_map(|key| {
+                key.column_indices
+                    .iter()
+                    .map(|old| old_to_new_col.get(old).cloned())
+                    .collect::<Option<Vec<_>>>()
+                    .and_then(|v| if v.is_empty() { None } else { Some(v) })
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/flow/src/repr/relation.rs:138]: > If I understand correctly, `collect()` here will return `None` when it encounters the first `None` of the iterator, so is `is_empty` in `and_then` here redundant?

This function is expected to remove empty keys like `column_indices` is empty, I should add this in comment

File: src/flow/src/transform.rs [hunk 4]
```diff
@@ -79,38 +92,32 @@ impl FunctionExtensions {
...
+        .context(InvalidQueryPlanSnafu)?;
+    let LogicalPlan::DfPlan(plan) = plan;
+
+    // encode then decode so to rely on the impl of conversion from logical plan to substrait plan
+    let bytes = DFLogicalSubstraitConvertor {}
+        .encode(&plan)
+        .context(InvalidQuerySubstraitSnafu)?;
+
+    let sub_plan = substrait::substrait_proto::proto::Plan::decode(bytes)
+        .map_err(|inner| InvalidQueryProstSnafu { inner }.build())?;
```
[the remark this answers is not in the document]
Comment (discord9) [reply src/flow/src/transform.rs:117]: > how about `datafusion_substrait::logical_plan::producer::to_substrait_plan` instead?

Refactored DFLogicalSubstraitConvertor a bit and did that

**Review by zhongzc — APPROVED** [review]

**Review by fengjiachun — APPROVED** [review]
LGTM

**Comment by discord9** [comment]
@waynexia do you want to take a look on those reexports newly added on common-substraits crate?

**Comment by waynexia** [comment]
They look good to me 👍

**Pull request merged**